plugins/obs-webrtc/whep-source.cpp

```diff
@@ -0,0 +1,511 @@
+	  running(false),
+	  start_stop_mutex(),
+	  start_stop_thread(),
+	  last_frame(std::chrono::system_clock::now()),
+	  last_audio_rtp_timestamp(0),
+	  last_video_rtp_timestamp(0),
+	  last_audio_pts(0),
+	  last_video_pts(0)
+{
+
+	this->video_av_codec_context = std::shared_ptr<AVCodecContext>(
+		this->CreateVideoAVCodecDecoder(),
+		[](AVCodecContext *ctx) { avcodec_free_context(&ctx); });
+
+	this->audio_av_codec_context = std::shared_ptr<AVCodecContext>(
+		this->CreateAudioAVCodecDecoder(),
+		[](AVCodecContext *ctx) { avcodec_free_context(&ctx); });
+
+	this->av_packet = std::shared_ptr<AVPacket>(
+		av_packet_alloc(), [](AVPacket *pkt) { av_packet_free(&pkt); });
+	this->av_frame =
+		std::shared_ptr<AVFrame>(av_frame_alloc(), [](AVFrame *frame) {
+			av_frame_free(&frame);
+		});
+
+	Update(settings);
+}
+
+WHEPSource::~WHEPSource()
+{
+	running = false;
+
+	Stop();
+
+	std::lock_guard<std::mutex> l(start_stop_mutex);
+	if (start_stop_thread.joinable())
+		start_stop_thread.join();
+}
+
+void WHEPSource::Stop()
+{
+	std::lock_guard<std::mutex> l(start_stop_mutex);
+	if (start_stop_thread.joinable())
+		start_stop_thread.join();
+
+	start_stop_thread = std::thread(&WHEPSource::StopThread, this);
+}
+
+void WHEPSource::StopThread()
+{
+	if (peer_connection != nullptr) {
+		peer_connection->close();
+		peer_connection = nullptr;
+		audio_track = nullptr;
+		video_track = nullptr;
+	}
+
+	SendDelete();
+}
+
+void WHEPSource::SendDelete()
+{
+	if (resource_url.empty()) {
+		do_log(LOG_DEBUG,
+		       "No resource URL available, not sending DELETE");
+		return;
+	}
+
+	char error_buffer[CURL_ERROR_SIZE] = {};
+	CURLcode curl_code;
+	auto status = send_delete(bearer_token, resource_url, &curl_code,
+				  error_buffer);
+	if (status == webrtc_network_status::Success) {
+		do_log(LOG_DEBUG,
+		       "Successfully performed DELETE request for resource URL");
+		resource_url.clear();
+	} else if (status == webrtc_network_status::DeleteFailed) {
+		do_log(LOG_WARNING,
+		       "DELETE request for resource URL failed. Reason: %s",
+		       error_buffer[0] ? error_buffer
+				       : curl_easy_strerror(curl_code));
+	} else if (status == webrtc_network_status::InvalidHTTPStatusCode) {
+		do_log(LOG_WARNING,
+		       "DELETE request for resource URL returned non-200 Status Code");
+	}
+}
+
+obs_properties_t *WHEPSource::GetProperties()
+{
+	obs_properties_t *ppts = obs_properties_create();
+
+	obs_properties_set_flags(ppts, OBS_PROPERTIES_DEFER_UPDATE);
+	obs_properties_add_text(ppts, "endpoint_url", "URL", OBS_TEXT_DEFAULT);
+	obs_properties_add_text(ppts, "bearer_token",
+				obs_module_text("Service.BearerToken"),
+				OBS_TEXT_PASSWORD);
+
+	return ppts;
+}
+
+void WHEPSource::Update(obs_data_t *settings)
+{
+	endpoint_url =
+		std::string(obs_data_get_string(settings, "endpoint_url"));
+	bearer_token =
+		std::string(obs_data_get_string(settings, "bearer_token"));
+
+	if (endpoint_url.empty() || bearer_token.empty()) {
```

Comment:
It seems WHEP Source always requires a Bearer token to be set, whereas WHIP Output does not have this requirement. 
Would it be possible to allow an empty Bearer token here?